docs/development-guide/extending-gaffer.md

```diff
@@ -0,0 +1,217 @@
+# Extending Gaffer
+
+Extending Gaffer can mean a few things to a user, this guide aims to cover
+general use cases for how and why you may want to extend the capabilities or add
+customisation to Gaffer.
+
+## Writing Custom Classes
+
+As Gaffer is Java based to create and load additional libraries you will first
+need to write your custom classes. Now these can be for a range of uses, Gaffer
+allows for many places to use custom classes such as, custom operations,
+aggregation functions, element generators etc.
+
+Depending on what type of class you are writing e.g. an `Operation` you may
+need extend or implement one of the interface classes already in Gaffer.
+Assuming you are using Maven for package management you can add Gaffer as
+a dependency to gain access to its classes like below:
+
+```xml
+<dependency>
+    <groupId>uk.gov.gchq.gaffer</groupId>
+    <artifactId>${gaffer.component}</artifactId>
+    <version>${gaffer.version}</version>
+</dependency>
+```
+
+!!! tip
+    See the [components breakdown](./project-structure/components/components.md)
+    for a list of key Gaffer modules.
+
+### Writing an Operation
+
+In Gaffer we use the term Operation to describe some component of a
+manipulation/query affecting a graph, these are chained together in `OperationChains`
+and this is how we execute manipulations/queries to our graphs. Whilst
+there are a wealth of existing operations and their associated handlers that are
+provided out of the box with Gaffer; in this section we'll talk about how you
+might go about adding a new operation/handler with some simple examples.
+
+#### The Operation Interface
+
+All Operations in one form or another must adhere to and implement the [`Operation`](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/operation/Operation.html)
+interface, this interface describes the building blocks an operation must be built from.
+
+One important distinction here is that an implementation of `Operation` should not
+contain the actual logic for an `Operation`, think of these as a task definition
+or a instruction manual for what an `Operation` should do. This then goes hand in
+hand with our `OperationHandler`(s) which define the actual functionality
+either generically or split into multiple handlers for different stores.
+
+#### Selecting an Interface
+
+In general most `Operation` implementations don't actually directly implement the
+`Operation` Interface, they actually implement one or more extended interfaces,
+for example the `Input` and `Output` interfaces extend on top of the core `Operation`
+interface and concordantly the `InputOutput` interface is a combination of these
+two to be used. These all live inside the Gaffer Core library under the
```

Comment:
```suggestion
two to be used. These all live inside the Gaffer Core library under the
```
The end of this sentence is a bit weird, could be phrased a little clearer than 'two to be used' I think